### PR TITLE
Feature/9.6/booking reservation request

### PR DESCRIPTION
**Pull request by vradovic**
This PR adds creating reservation requests.

### CODE REVIEW

File: src/main/java/com/lunark/lunark/reservations/service/ReservationService.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+        Optional<Account> guest = accountRepository.findByEmail(username);
+        if (guest.isEmpty()) {
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "user does not exist");
+        }
+
+        double totalPrice = calculatePrice(
+                reservationDto.getStartDate(),
+                reservationDto.getEndDate(),
+                property.get().getAvailabilityEntries().stream().toList()
```
Comment (MomirMilutinovic) [inline src/main/java/com/lunark/lunark/reservations/service/ReservationService.java:59]: I suggest shortening this line by removing the .stream.toList() and changing the type of this parameter to Collection. `Collection`s support the same stream operation as `List`s. 
```suggestion
                property.get().getAvailabilityEntries()
```

File: src/main/java/com/lunark/lunark/reservations/dto/ReservationResponseDto.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import java.time.LocalDate;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class ReservationResponseDto {
+    PropertyResponseDto property;
```
Comment (MomirMilutinovic) [inline src/main/java/com/lunark/lunark/reservations/dto/ReservationResponseDto.java:15]: Do we need to return the entire property when creating a reservation?
```suggestion
```

File: src/main/java/com/lunark/lunark/reservations/controller/ReservationController.java [hunk 3]
```diff
@@ -15,10 +23,24 @@
...
+    public ReservationController(IReservationService reservationService, ModelMapper modelMapper) {
+        this.reservationService = reservationService;
+        this.modelMapper = modelMapper;
+    }
+
     @PostMapping(path = "", consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
-    public ResponseEntity<ReservationConfirmationDto> createReservation(@RequestHeader("x-access-token") String guestToken, @RequestBody ReservationRequestDTO reservationRequestDTO) {
-        ReservationConfirmationDto reservationConfirmationDto = new ReservationConfirmationDto("Vila Golija", LocalDate.of(2024, 6, 14), LocalDate.of(2023, 6, 16), 9959, 3);
-        return new ResponseEntity<>(reservationConfirmationDto, HttpStatus.CREATED);
+    @PreAuthorize("hasAuthority('guest')")
```
Comment (MomirMilutinovic) [inline src/main/java/com/lunark/lunark/reservations/controller/ReservationController.java:36]: Did you mean to put this in uppercase? I get 403 errors if it's left as is.
```suggestion
    @PreAuthorize("hasAuthority('GUEST')")
```

File: src/main/java/com/lunark/lunark/reservations/service/ReservationService.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+                .status(property.get().isAutoApproveEnabled() ? ReservationStatus.ACCEPTED : ReservationStatus.PENDING)
+                .numberOfGuests(reservationDto.getNumberOfGuests())
+                .startDate(reservationDto.getStartDate())
+                .endDate(reservationDto.getEndDate())
+                .build();
+
+        return this.reservationRepository.save(reservation);
+    }
+
+    private double calculatePrice(LocalDate start, LocalDate end, List<PropertyAvailabilityEntry> entries) {
```
Comment (MomirMilutinovic) [inline src/main/java/com/lunark/lunark/reservations/service/ReservationService.java:75]:
```suggestion
    private double calculatePrice(LocalDate start, LocalDate end, Collection<PropertyAvailabilityEntry> entries) {
```

File: src/main/java/com/lunark/lunark/reservations/service/ReservationService.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+import com.lunark.lunark.reservations.repository.IReservationRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.stereotype.Service;
+import org.springframework.web.server.ResponseStatusException;
+
+import java.security.Principal;
+import java.time.LocalDate;
+import java.util.List;
```
Comment (MomirMilutinovic) [inline src/main/java/com/lunark/lunark/reservations/service/ReservationService.java:20]:
```suggestion
import java.util.List;
import java.util.Collection;
```

**Review by MomirMilutinovic — CHANGES_REQUESTED** [review]
Great work.
I found a way to make this PR a bit cleaner and a small typo.

File: src/main/java/com/lunark/lunark/reservations/dto/ReservationResponseDto.java (hunk 2 above)
Comment (vradovic) [reply src/main/java/com/lunark/lunark/reservations/dto/ReservationResponseDto.java:15]: I would leave this as when the guest browses his reservations, we would display property details as well.

**Review by duolok — APPROVED** [review]
LGTM

**Pull request merged**